File: src/Java/myservicebus/src/main/java/com/myservicebus/choreography/ChoreographyBuilder.java
package com.myservicebus.choreography;

import com.myservicebus.MessageUrn;

import java.time.Duration;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Objects;
import java.util.function.Consumer;

/** Builds a normalized declaration of application-owned choreography reactions. */
public final class ChoreographyBuilder {
    private final String choreographyId;
    private final String definitionVersion;
    private final String owner;
    private final List<ChoreographyStep> steps = new ArrayList<>();

    public ChoreographyBuilder(String choreographyId, String definitionVersion, String owner) {
        this.choreographyId = required(choreographyId, "choreographyId");
        this.definitionVersion = required(definitionVersion, "definitionVersion");
        this.owner = required(owner, "owner");
    }

    public ChoreographyBuilder step(
            String id,
            Class<?> triggerMessageType,
            Consumer<ChoreographyStepBuilder> configure) {
        Objects.requireNonNull(triggerMessageType, "triggerMessageType");
        return step(id, MessageUrn.forClass(triggerMessageType), configure);
    }

    public ChoreographyBuilder step(
            String id,
            String triggerMessageUrn,
            Consumer<ChoreographyStepBuilder> configure) {
        Objects.requireNonNull(configure, "configure");

        String stepId = required(id, "id");
        if (steps.stream().anyMatch(step -> step.id().equals(stepId))) {
            throw new IllegalArgumentException("A choreography step with ID '" + stepId + "' is already declared.");
        }

        ChoreographyStepBuilder builder = new ChoreographyStepBuilder(
                stepId,
                required(triggerMessageUrn, "triggerMessageUrn"));
        configure.accept(builder);
        steps.add(builder.build());
        return this;
    }

    public ChoreographyFragment build() {
        if (steps.isEmpty()) {
            throw new IllegalStateException("A choreography fragment must declare at least one step.");
        }

        List<ChoreographyStep> normalizedSteps = steps.stream()
                .sorted(Comparator.comparing(ChoreographyStep::id))
                .toList();
        return new ChoreographyFragment(
                ChoreographyFragment.CURRENT_SCHEMA_VERSION,
                choreographyId,
                definitionVersion,
                owner,
                normalizedSteps);
    }

    static String required(String value, String parameterName) {
        if (value == null || value.isBlank()) {
            throw new IllegalArgumentException(parameterName + " cannot be empty or whitespace.");
        }
        return value;
    }

    public static final class ChoreographyStepBuilder {
        private static final Comparator<String> NULLABLE_STRING_COMPARATOR =
                Comparator.nullsFirst(Comparator.naturalOrder());

        private final String id;
        private final String triggerMessageUrn;
        private final List<ChoreographyOutput> outputs = new ArrayList<>();
        private String ownerComponent;

        private ChoreographyStepBuilder(String id, String triggerMessageUrn) {
            this.id = id;
            this.triggerMessageUrn = triggerMessageUrn;
        }

        public ChoreographyStepBuilder ownedBy(Class<?> componentType) {
            Objects.requireNonNull(componentType, "componentType");
            return ownedBy(componentType.getName());
        }

        public ChoreographyStepBuilder ownedBy(String component) {
            ownerComponent = required(component, "component");
            return this;
        }

        public ChoreographyStepBuilder sends(Class<?> messageType, String destination) {
            return sends(messageType, destination, null);
        }

        public ChoreographyStepBuilder sends(
                Class<?> messageType,
                String destination,
                Consumer<ChoreographyOutputBuilder> configure) {
            Objects.requireNonNull(messageType, "messageType");
            return sends(MessageUrn.forClass(messageType), destination, configure);
        }

        public ChoreographyStepBuilder sends(String messageUrn, String destination) {
            return sends(messageUrn, destination, null);
        }

        public ChoreographyStepBuilder sends(
                String messageUrn,
                String destination,
                Consumer<ChoreographyOutputBuilder> configure) {
            return add(
                    ChoreographyOperationKind.SEND,
                    messageUrn,
                    required(destination, "destination"),
                    configure);
        }

        public ChoreographyStepBuilder publishes(Class<?> messageType) {
            return publishes(messageType, null);
        }

        public ChoreographyStepBuilder publishes(
                Class<?> messageType,
                Consumer<ChoreographyOutputBuilder> configure) {
            Objects.requireNonNull(messageType, "messageType");
            return publishes(MessageUrn.forClass(messageType), configure);
        }

        public ChoreographyStepBuilder publishes(String messageUrn) {
            return publishes(messageUrn, null);
        }

        public ChoreographyStepBuilder publishes(
                String messageUrn,
                Consumer<ChoreographyOutputBuilder> configure) {
            return add(ChoreographyOperationKind.PUBLISH, messageUrn, null, configure);
        }

        public ChoreographyStepBuilder responds(Class<?> messageType) {
            return responds(messageType, null);
        }

        public ChoreographyStepBuilder responds(
                Class<?> messageType,
                Consumer<ChoreographyOutputBuilder> configure) {
            Objects.requireNonNull(messageType, "messageType");
            return responds(MessageUrn.forClass(messageType), configure);
        }

        public ChoreographyStepBuilder responds(String messageUrn) {
            return responds(messageUrn, null);
        }

        public ChoreographyStepBuilder responds(
                String messageUrn,
                Consumer<ChoreographyOutputBuilder> configure) {
            return add(ChoreographyOperationKind.RESPOND, messageUrn, null, configure);
        }

        public ChoreographyStepBuilder schedules(Class<?> messageType) {
            return schedules(messageType, null, null);
        }

        public ChoreographyStepBuilder schedules(
                Class<?> messageType,
                Consumer<ChoreographyOutputBuilder> configure) {
            return schedules(messageType, null, configure);
        }

        public ChoreographyStepBuilder schedules(
                Class<?> messageType,
                String destination,
                Consumer<ChoreographyOutputBuilder> configure) {
            Objects.requireNonNull(messageType, "messageType");
            return schedules(MessageUrn.forClass(messageType), destination, configure);
        }

        public ChoreographyStepBuilder schedules(String messageUrn) {
            return schedules(messageUrn, null, null);
        }

        public ChoreographyStepBuilder schedules(
                String messageUrn,
                Consumer<ChoreographyOutputBuilder> configure) {
            return schedules(messageUrn, null, configure);
        }

        public ChoreographyStepBuilder schedules(
                String messageUrn,
                String destination,
                Consumer<ChoreographyOutputBuilder> configure) {
            return add(ChoreographyOperationKind.SCHEDULE, messageUrn, destination, configure);
        }

        public ChoreographyStepBuilder terminates() {
            return terminates(null);
        }

        public ChoreographyStepBuilder terminates(Consumer<ChoreographyOutputBuilder> configure) {
            return add(ChoreographyOperationKind.TERMINAL, null, null, configure);
        }

        private ChoreographyStep build() {
            if (outputs.isEmpty()) {
                throw new IllegalStateException(
                        "Choreography step '" + id + "' must declare at least one output or terminal outcome.");
            }

            List<ChoreographyOutput> normalizedOutputs = outputs.stream()
                    .sorted(Comparator.comparing(ChoreographyOutput::kind)
                            .thenComparing(ChoreographyOutput::messageUrn, NULLABLE_STRING_COMPARATOR)
                            .thenComparing(ChoreographyOutput::destination, NULLABLE_STRING_COMPARATOR)
                            .thenComparing(ChoreographyOutput::requirement)
                            .thenComparing(ChoreographyOutput::minCount, Comparator.nullsFirst(Comparator.naturalOrder()))
                            .thenComparing(ChoreographyOutput::maxCount, Comparator.nullsFirst(Comparator.naturalOrder()))
                            .thenComparing(
                                    ChoreographyOutput::withinMilliseconds,
                                    Comparator.nullsFirst(Comparator.naturalOrder())))
                    .toList();
            return new ChoreographyStep(id, triggerMessageUrn, ownerComponent, normalizedOutputs);
        }

        private ChoreographyStepBuilder add(
                ChoreographyOperationKind kind,
                String messageUrn,
                String destination,
                Consumer<ChoreographyOutputBuilder> configure) {
            if (kind != ChoreographyOperationKind.TERMINAL) {
                messageUrn = required(messageUrn, "messageUrn");
            }

            ChoreographyOutputBuilder builder = new ChoreographyOutputBuilder(kind, messageUrn, destination);
            if (configure != null) {
                configure.accept(builder);
            }
            outputs.add(builder.build());
            return this;
        }
    }

    public static final class ChoreographyOutputBuilder {
        private final ChoreographyOperationKind kind;
        private final String messageUrn;
        private final String destination;
        private ChoreographyRequirement requirement = ChoreographyRequirement.EXPECTED;
        private Integer minCount;
        private Integer maxCount;
        private Long withinMilliseconds;

        private ChoreographyOutputBuilder(
                ChoreographyOperationKind kind,
                String messageUrn,
                String destination) {
            this.kind = kind;
            this.messageUrn = messageUrn;
            this.destination = destination;
        }

        public ChoreographyOutputBuilder informational() {
            requirement = ChoreographyRequirement.INFORMATIONAL;
            return this;
        }

        public ChoreographyOutputBuilder optional() {
            requirement = ChoreographyRequirement.OPTIONAL;
            return this;
        }

        public ChoreographyOutputBuilder expected() {
            requirement = ChoreographyRequirement.EXPECTED;
            return this;
        }

        public ChoreographyOutputBuilder atLeast(int count) {
            minCount = nonNegative(count, "count");
            return this;
        }

        public ChoreographyOutputBuilder atMost(int count) {
            maxCount = nonNegative(count, "count");
            return this;
        }

        public ChoreographyOutputBuilder exactly(int count) {
            minCount = maxCount = nonNegative(count, "count");
            return this;
        }

        public ChoreographyOutputBuilder within(Duration duration) {
            Objects.requireNonNull(duration, "duration");
            long milliseconds = duration.toMillis();
            if (milliseconds <= 0) {
                throw new IllegalArgumentException("The time expectation must be at least one millisecond.");
            }
            withinMilliseconds = milliseconds;
            return this;
        }

        private ChoreographyOutput build() {
            if (minCount != null && maxCount != null && minCount > maxCount) {
                throw new IllegalStateException("The minimum output count cannot exceed the maximum output count.");
            }
            if (kind == ChoreographyOperationKind.TERMINAL
                    && (minCount != null || maxCount != null || withinMilliseconds != null)) {
                throw new IllegalStateException(
                        "A terminal outcome cannot declare output count or timing expectations.");
            }
            return new ChoreographyOutput(
                    kind,
                    messageUrn,
                    destination,
                    requirement,
                    minCount,
                    maxCount,
                    withinMilliseconds);
        }

        private static int nonNegative(int count, String parameterName) {
            if (count < 0) {
                throw new IllegalArgumentException(parameterName + " cannot be negative.");
            }
            return count;
        }
    }
}

File: src/MyServiceBus/Choreography/ChoreographyBuilder.cs
namespace MyServiceBus.Choreography;

/// <summary>
/// Builds a normalized declaration of application-owned choreography reactions.
/// </summary>
public sealed class ChoreographyBuilder
{
    private readonly string choreographyId;
    private readonly string definitionVersion;
    private readonly string owner;
    private readonly List<ChoreographyStep> steps = new();

    public ChoreographyBuilder(string choreographyId, string definitionVersion, string owner)
    {
        this.choreographyId = Required(choreographyId, nameof(choreographyId));
        this.definitionVersion = Required(definitionVersion, nameof(definitionVersion));
        this.owner = Required(owner, nameof(owner));
    }

    public ChoreographyBuilder Step<TTrigger>(string id, Action<ChoreographyStepBuilder> configure)
        => Step(id, MessageUrn.For(typeof(TTrigger)), configure);

    public ChoreographyBuilder Step(
        string id,
        string triggerMessageUrn,
        Action<ChoreographyStepBuilder> configure)
    {
        ArgumentNullException.ThrowIfNull(configure);

        var stepId = Required(id, nameof(id));
        if (steps.Any(step => string.Equals(step.Id, stepId, StringComparison.Ordinal)))
        {
            throw new ArgumentException($"A choreography step with ID '{stepId}' is already declared.", nameof(id));
        }

        var builder = new ChoreographyStepBuilder(stepId, Required(triggerMessageUrn, nameof(triggerMessageUrn)));
        configure(builder);
        steps.Add(builder.Build());
        return this;
    }

    public ChoreographyFragment Build()
    {
        if (steps.Count == 0)
        {
            throw new InvalidOperationException("A choreography fragment must declare at least one step.");
        }

        var fragment = new ChoreographyFragment(
            ChoreographyFragment.CurrentSchemaVersion,
            choreographyId,
            definitionVersion,
            owner,
            steps.OrderBy(step => step.Id, StringComparer.Ordinal).ToArray());
        fragment.Validate();
        return fragment;
    }

    internal static string Required(string value, string parameterName)
    {
        if (string.IsNullOrWhiteSpace(value))
        {
            throw new ArgumentException("The value cannot be empty or whitespace.", parameterName);
        }

        return value;
    }
}

public sealed class ChoreographyStepBuilder
{
    private readonly string id;
    private readonly string triggerMessageUrn;
    private readonly List<ChoreographyOutput> outputs = new();
    private string? ownerComponent;

    internal ChoreographyStepBuilder(string id, string triggerMessageUrn)
    {
        this.id = id;
        this.triggerMessageUrn = triggerMessageUrn;
    }

    public ChoreographyStepBuilder OwnedBy<TComponent>() => OwnedBy(typeof(TComponent).FullName ?? typeof(TComponent).Name);

    public ChoreographyStepBuilder OwnedBy(string component)
    {
        ownerComponent = ChoreographyBuilder.Required(component, nameof(component));
        return this;
    }

    public ChoreographyStepBuilder Sends<TMessage>(
        string destination,
        Action<ChoreographyOutputBuilder>? configure = null)
        => Sends(MessageUrn.For(typeof(TMessage)), destination, configure);

    public ChoreographyStepBuilder Sends(
        string messageUrn,
        string destination,
        Action<ChoreographyOutputBuilder>? configure = null)
        => Add(ChoreographyOperationKind.Send, messageUrn, ChoreographyBuilder.Required(destination, nameof(destination)), configure);

    public ChoreographyStepBuilder Publishes<TMessage>(Action<ChoreographyOutputBuilder>? configure = null)
        => Publishes(MessageUrn.For(typeof(TMessage)), configure);

    public ChoreographyStepBuilder Publishes(
        string messageUrn,
        Action<ChoreographyOutputBuilder>? configure = null)
        => Add(ChoreographyOperationKind.Publish, messageUrn, null, configure);

    public ChoreographyStepBuilder Responds<TMessage>(Action<ChoreographyOutputBuilder>? configure = null)
        => Responds(MessageUrn.For(typeof(TMessage)), configure);

    public ChoreographyStepBuilder Responds(
        string messageUrn,
        Action<ChoreographyOutputBuilder>? configure = null)
        => Add(ChoreographyOperationKind.Respond, messageUrn, null, configure);

    public ChoreographyStepBuilder Schedules<TMessage>(Action<ChoreographyOutputBuilder>? configure = null)
        => Schedules(MessageUrn.For(typeof(TMessage)), null, configure);

    public ChoreographyStepBuilder Schedules<TMessage>(
        string destination,
        Action<ChoreographyOutputBuilder>? configure = null)
        => Schedules(MessageUrn.For(typeof(TMessage)), ChoreographyBuilder.Required(destination, nameof(destination)), configure);

    public ChoreographyStepBuilder Schedules(
        string messageUrn,
        string? destination = null,
        Action<ChoreographyOutputBuilder>? configure = null)
        => Add(ChoreographyOperationKind.Schedule, messageUrn, destination, configure);

    public ChoreographyStepBuilder Terminates(Action<ChoreographyOutputBuilder>? configure = null)
        => Add(ChoreographyOperationKind.Terminal, null, null, configure);

    internal ChoreographyStep Build()
    {
        if (outputs.Count == 0)
        {
            throw new InvalidOperationException($"Choreography step '{id}' must declare at least one output or terminal outcome.");
        }

        var normalizedOutputs = outputs
            .OrderBy(output => output.Kind)
            .ThenBy(output => output.MessageUrn, StringComparer.Ordinal)
            .ThenBy(output => output.Destination, StringComparer.Ordinal)
            .ThenBy(output => output.Requirement)
            .ThenBy(output => output.MinCount)
            .ThenBy(output => output.MaxCount)
            .ThenBy(output => output.WithinMilliseconds)
            .ToArray();

        return new ChoreographyStep(id, triggerMessageUrn, ownerComponent, normalizedOutputs);
    }

    private ChoreographyStepBuilder Add(
        ChoreographyOperationKind kind,
        string? messageUrn,
        string? destination,
        Action<ChoreographyOutputBuilder>? configure)
    {
        if (kind != ChoreographyOperationKind.Terminal)
        {
            messageUrn = ChoreographyBuilder.Required(messageUrn!, nameof(messageUrn));
        }

        var builder = new ChoreographyOutputBuilder(kind, messageUrn, destination);
        configure?.Invoke(builder);
        outputs.Add(builder.Build());
        return this;
    }
}

public sealed class ChoreographyOutputBuilder
{
    private readonly ChoreographyOperationKind kind;
    private readonly string? messageUrn;
    private readonly string? destination;
    private ChoreographyRequirement requirement = ChoreographyRequirement.Expected;
    private int? minCount;
    private int? maxCount;
    private long? withinMilliseconds;

    internal ChoreographyOutputBuilder(ChoreographyOperationKind kind, string? messageUrn, string? destination)
    {
        this.kind = kind;
        this.messageUrn = messageUrn;
        this.destination = destination;
    }

    public ChoreographyOutputBuilder Informational()
    {
        requirement = ChoreographyRequirement.Informational;
        return this;
    }

    public ChoreographyOutputBuilder Optional()
    {
        requirement = ChoreographyRequirement.Optional;
        return this;
    }

    public ChoreographyOutputBuilder Expected()
    {
        requirement = ChoreographyRequirement.Expected;
        return this;
    }

    public ChoreographyOutputBuilder AtLeast(int count)
    {
        minCount = NonNegative(count, nameof(count));
        return this;
    }

    public ChoreographyOutputBuilder AtMost(int count)
    {
        maxCount = NonNegative(count, nameof(count));
        return this;
    }

    public ChoreographyOutputBuilder Exactly(int count)
    {
        minCount = maxCount = NonNegative(count, nameof(count));
        return this;
    }

    public ChoreographyOutputBuilder Within(TimeSpan duration)
    {
        var milliseconds = checked((long)duration.TotalMilliseconds);
        if (milliseconds <= 0)
        {
            throw new ArgumentOutOfRangeException(nameof(duration), "The time expectation must be at least one millisecond.");
        }

        withinMilliseconds = milliseconds;
        return this;
    }

    internal ChoreographyOutput Build()
    {
        if (minCount > maxCount)
        {
            throw new InvalidOperationException("The minimum output count cannot exceed the maximum output count.");
        }

        if (kind == ChoreographyOperationKind.Terminal &&
            (minCount is not null || maxCount is not null || withinMilliseconds is not null))
        {
            throw new InvalidOperationException("A terminal outcome cannot declare output count or timing expectations.");
        }

        return new ChoreographyOutput(
            kind,
            messageUrn,
            destination,
            requirement,
            minCount,
            maxCount,
            withinMilliseconds);
    }

    private static int NonNegative(int count, string parameterName)
    {
        if (count < 0)
        {
            throw new ArgumentOutOfRangeException(parameterName, "The output count cannot be negative.");
        }

        return count;
    }
}
